components/gitpod-cli/cmd/validate.go

```diff
@@ -564,59 +566,85 @@ func pipeTask(ctx context.Context, task *api.TaskStatus, supervisor *supervisor.
 	}
 }
 
-func listenTerminal(ctx context.Context, task *api.TaskStatus, supervisor *supervisor.SupervisorClient, runLog *logrus.Entry) error {
-	listen, err := supervisor.Terminal.Listen(ctx, &api.ListenTerminalRequest{
-		Alias: task.Terminal,
-	})
-	if err != nil {
-		return err
-	}
+// TerminalReader is an interface for anything that can receive terminal data (this is abstracted for use in testing)
+type TerminalReader interface {
+	Recv() ([]byte, error)
+}
 
-	pr, pw := io.Pipe()
-	defer pr.Close()
-	defer pw.Close()
+type LinePrinter func(string)
 
-	scanner := bufio.NewScanner(pr)
-	const maxTokenSize = 1 * 1024 * 1024 // 1 MB
-	buf := make([]byte, maxTokenSize)
-	scanner.Buffer(buf, maxTokenSize)
+// processTerminalOutput reads from a TerminalReader, processes the output, and calls the provided LinePrinter for each complete line.
+// It handles UTF-8 decoding of characters split across chunks and control characters (\n \r \b).
+func processTerminalOutput(reader TerminalReader, printLine LinePrinter) error {
+	var buffer, line bytes.Buffer
 
-	go func() {
-		defer pw.Close()
-		for {
-			resp, err := listen.Recv()
-			if err != nil {
-				_ = pw.CloseWithError(err)
-				return
-			}
+	flushLine := func() {
+		if line.Len() > 0 {
```

Comment:
this has a catch.

Pros:
- When we read Windows-like line endings (`\r\n`), we treat it as a single line-break

Cons:
- A hundred line breaks will be treated as 0 if the line has no other content